python_modules/dagster/dagster/core/execution/context/system.py

```diff
@@ -548,6 +550,68 @@ def step_output_capture(self) -> Dict[StepOutputHandle, Any]:
     def previous_attempt_count(self) -> int:
         return self._previous_attempt_count
 
+    def has_asset_partitions_for_input(self, input_name: str) -> bool:
+        tags = self._plan_data.pipeline_run.tags
+
+        if "asset_partitions" in tags:
```

Comment:
Yeah, having the tag value be some json serialized string is starting to feel a little bad to me.  This is because we can have arbitrarily complex combinations of asset partitions (e.g. N different partition names where N is the number of assets in the job?)?  

I was hoping we could consolidate this information just by the existing `partition_key` information (which is now the "job-level partition key"), and then forcing the mapping for each asset partition from that partition key.

I know this is how we've constrained the user-facing API, but I think this would also simplify the internal API as well.

The thing about subclassing the IOManager / silo-ing the more complex behavior into asset-partition specific implementations that resonates with me is that the APIs make more sense then more complex you want to go.  As it is, looking at `asset_partition_key_range_for_input` as an available method on the context is noise that I have to ignore.  

Since we're constructing SDA jobs the same way (using `build_assets_job`) is it a heavy lift to use a specialized `AssetsIOManager` that has access to the definitions?  That does feel a lot cleaner to me.

Also, I'm not an IOManager expert.  Is there a reason we would need to split out a special `handle_asset_output` method, instead of just `handle_output`?

Reply:
> I was hoping we could consolidate this information just by the existing partition_key information (which is now the "job-level partition key"), and then forcing the mapping for each asset partition from that partition key.

Hmm yes I think we could do this - I'll try it out.

> Since we're constructing SDA jobs the same way (using build_assets_job) is it a heavy lift to use a specialized AssetsIOManager that has access to the definitions? That does feel a lot cleaner to me.

I don't think it's a heavy lift per se, but it yet adds one more thing for users to get wrong. I can imagine a series of #dagster-support answers telling users that the reason they're getting X error is because they forgot to extend the right IOManager class.

FWIW, our context objects are already a grab bag of a bunch of different methods and properties, only a small subset of which get used in any particular program. So I don't think it's especially distracting to add more.

> Also, I'm not an IOManager expert. Is there a reason we would need to split out a special handle_asset_output method, instead of just handle_output?

The way I had imagined this would work is that `AssetIOManager` would extend `IOManager`, and its `handle_output` implementation would build an `AssetOutputContext` out of the `OutputContext` provided to it, and then invoke a user-overridden `handle_asset_output` method. I.e. the implementation of `handle_output` for an `AssetIOManager` needs to include the framework code that wraps the `OutputContext`, so it can't also be a user-supplied method.

Very possible I'm missing an alternative approach that sidesteps that issue though?